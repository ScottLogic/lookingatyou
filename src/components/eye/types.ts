export interface IInnerPartProps {
    transitionStyle: React.CSSProperties;
    pupilRadius: number;
    skewTransform: string;
    groupProps: {
        className: string;
        style: {
            transition: string;
        };
        transform: string;
    };
}
