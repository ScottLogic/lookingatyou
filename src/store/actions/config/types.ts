import {
    architectures,
    inputResolutions,
    multipliers,
    outputStrides,
} from '../../../models/objectDetection';

export enum ConfigSetAction {
    APP = 'SET_APP_CONFIG',
    ADVANCED = 'SET_ADVANCED',
    MODEL = 'SET_MODEL_CONFIG',
    DETECTION = 'SET_DETECTION_CONFIG',
    RESET = 'RESET_CONFIG',
}

export interface IConfigState {
    advancedConfig: IAdvancedConfig;
    appConfig: IAppConfig;
}

export type PartialConfig =
    | Partial<IAppConfig>
    | Partial<IModelStringConfig>
    | Partial<IModelConfig>
    | Partial<IDetectionConfig>;

export interface IAppConfig {
    xSensitivity: number;
    ySensitivity: number;
    fps: number;
    toggleAdvanced: boolean;
    irisColor: string;
    showHelp: boolean;
}

export interface IAdvancedConfig {
    toggleReflection: boolean;
    reflectionOpacity: number;
    toggleDebug: boolean;
    modelConfig: IModelConfig;
    detectionConfig: IDetectionConfig;
}

export interface IModelConfig {
    architecture: architectures;
    outputStride: outputStrides;
    multiplier: multipliers;
    inputResolution: inputResolutions;
}

export interface IModelStringConfig {
    architecture: architectures;
    outputStride: string;
    multiplier: string;
    inputResolution: string;
}

export interface IDetectionConfig {
    flipHorizontal: boolean;
    maxDetections: number;
    scoreThreshold: number;
    nmsRadius: number;
}

export interface ISetAppConfigAction {
    readonly type: ConfigSetAction.APP;
    readonly payload: Partial<IAppConfig>;
}

export interface ISetModelConfigAction {
    readonly type: ConfigSetAction.MODEL;
    readonly payload: Partial<IModelConfig>;
}

export interface ISetDetectionConfigAction {
    readonly type: ConfigSetAction.DETECTION;
    readonly payload: Partial<IDetectionConfig>;
}

export interface ISetAdvancedConfigAction {
    readonly type: ConfigSetAction.ADVANCED;
    readonly payload: Partial<IAdvancedConfig>;
}

export interface IResetConfigAction {
    readonly type: ConfigSetAction.RESET;
}

export type ConfigAction =
    | ISetAppConfigAction
    | ISetAdvancedConfigAction
    | ISetModelConfigAction
    | ISetDetectionConfigAction
    | IResetConfigAction;
