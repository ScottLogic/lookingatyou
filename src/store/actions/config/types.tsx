import IUserConfig from '../../../components/configMenu/IUserConfig';
export const UPDATE_CONFIG = 'UPDATE_CONFIG';
export const RESET_CONFIG = 'RESET_CONFIG';
export interface IConfigState {
    config: IUserConfig;
}
export interface ISetConfigPayload {
    partialConfig: Partial<IUserConfig>;
}
interface IUpdateConfigAction {
    type: typeof UPDATE_CONFIG;
    payload: ISetConfigPayload;
}
interface IResetConfigAction {
    type: typeof RESET_CONFIG;
}
export type ConfigActionTypes = IUpdateConfigAction | IResetConfigAction;
