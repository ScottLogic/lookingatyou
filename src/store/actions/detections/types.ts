import { IDetections } from '../../../models/objectDetection';
import { ITargets } from '../../../utils/types';

export const SET_MODEL_LOADED = 'SET_MODEL_LOADED';
export const SET_TARGETS = 'SET_TARGETS';
export const SET_DETECTIONS = 'SET_DETECTIONS';
export const SET_BRIGHT = 'SET_BRIGHT';
export const SET_DILATION = 'SET_DILATION';
export const SET_PERSON = 'SET_PERSON';
export const SET_LEFT = 'SET_LEFT';
export const SET_OPEN = 'SET_OPEN';
export const SET_SQUINT = 'SET_SQUINT';

export interface IDetectionState {
    isModelLoaded: boolean;
    tooBright: boolean;
    left: boolean;
    dilationCoefficient: number;
    eyesOpenCoefficient: number;
    personDetected: boolean;
    isSquinting: boolean;
    target: ITargets;
    detections: IDetections;
}

export interface ISetLoadedAction {
    type: 'SET_MODEL_LOADED';
    payload: boolean;
}

export interface ISetTargetAction {
    type: 'SET_TARGETS';
    payload: ITargets;
}

export interface ISetDetectionsAction {
    type: 'SET_DETECTIONS';
    payload: IDetections;
}

export interface ISetBrightAction {
    type: 'SET_BRIGHT';
    payload: boolean;
}

export interface ISetLeftAction {
    type: 'SET_LEFT';
    payload: boolean;
}

export interface ISetDilationAction {
    type: 'SET_DILATION';
    payload: number;
}

export interface ISetOpenAction {
    type: 'SET_OPEN';
    payload: number;
}

export interface ISetPersonAction {
    type: 'SET_PERSON';
    payload: boolean;
}

export interface ISetSquintAction {
    type: 'SET_SQUINT';
    payload: boolean;
}

export type DetectionActionType =
    | ISetLoadedAction
    | ISetTargetAction
    | ISetDetectionsAction
    | ISetBrightAction
    | ISetDilationAction
    | ISetLeftAction
    | ISetOpenAction
    | ISetPersonAction
    | ISetSquintAction;
