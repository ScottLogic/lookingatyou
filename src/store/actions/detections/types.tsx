import { IDetection } from '../../../models/objectDetection';
import { ICoords } from '../../../utils/types';

export const SET_MODEL_LOADED = 'SET_MODEL_LOADED';
export const SET_TARGET = 'SET_TARGET';
export const SET_DETECTIONS = 'SET_DETECTIONS';

export interface IDetectionState {
    isModelLoaded: boolean;
    target: ICoords;
    detections: IDetection[];
}

export interface ISetLoadedAction {
    type: 'SET_MODEL_LOADED';
    payload: boolean;
}

export interface ISetTargetAction {
    type: 'SET_TARGET';
    payload: ICoords;
}

export interface ISetDetectionsAction {
    type: 'SET_DETECTIONS';
    payload: IDetection[];
}

export type DetectionActionType =
    | ISetLoadedAction
    | ISetTargetAction
    | ISetDetectionsAction;
