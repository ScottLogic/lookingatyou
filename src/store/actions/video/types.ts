export const SET_VIDEO = 'SET_VIDEO';
export const SET_VIDEO_STREAMS = 'SET_VIDEO_STREAMS';
export const TOGGLE_WEBCAM_AVAILABLE = 'TOGGLE_WEBCAM_AVAILABLE';
export const SET_IMAGE_DATA = 'SET_IMAGE_DATA';

export interface IVideo {
    deviceId: string;
    video?: HTMLVideoElement | undefined;
    width: number;
    height: number;
    stream: MediaStream | undefined;
}

export interface IVideoState {
    webcamAvailable: boolean;
    videos: { [deviceId: string]: IVideo };
    image: ImageData;
}

export interface ISetVideoPayload {
    video: HTMLVideoElement;
    deviceId: string;
}

interface ISetVideoAction {
    type: typeof SET_VIDEO;
    payload: ISetVideoPayload;
}

interface ISetVideoStreamsAction {
    type: typeof SET_VIDEO_STREAMS;
    videos: { [deviceId: string]: IVideo };
}

interface IToggleWebcamAvailable {
    type: typeof TOGGLE_WEBCAM_AVAILABLE;
}

interface ISetImageDataAction {
    type: typeof SET_IMAGE_DATA;
    image: ImageData;
}

export type VideoActionTypes =
    | ISetVideoAction
    | ISetVideoStreamsAction
    | IToggleWebcamAvailable
    | ISetImageDataAction;
