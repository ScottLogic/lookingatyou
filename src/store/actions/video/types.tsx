export const SET_VIDEO = 'SET_VIDEO';
export const SET_VIDEO_STREAMS = 'SET_VIDEO_STREAMS';
export const SET_DIMENSIONS = 'SET_DIMENSIONS';

export interface IVideo {
  deviceId: string,
  video?: HTMLVideoElement | undefined,
  width: number,
  height: number,
  stream: MediaStream | undefined,
}

export interface IVideoState {
  videos: IVideo[],
}

export interface IDimensions {
  width: number,
  height: number,
}

interface ISetVideoAction {
  type: typeof SET_VIDEO,
  video: HTMLVideoElement,
}

interface ISetDimensionsAction {
  type: typeof SET_DIMENSIONS,
  dimensions: IDimensions,
}

interface ISetVideoStreamsAction {
  type: typeof SET_VIDEO_STREAMS,
  videos: IVideo[],
}

export type VideoActionTypes = ISetVideoAction | ISetDimensionsAction | ISetVideoStreamsAction;
