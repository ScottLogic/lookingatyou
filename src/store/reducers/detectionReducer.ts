import { eyelidPosition, maxNumTargetsToConsider } from '../../AppConstants';
import { IObjectDetector } from '../../models/objectDetection';
import {
    calculateColourMatch,
    IColour,
} from '../../utils/objectSelection/select';
import { ICoords } from '../../utils/types';
import {
    DetectionActionType,
    IDetectionState,
    ISetDetectionsActionPayload,
    SET_DETECTIONS,
    SET_IDLE_TARGET,
    SET_INTERVAL,
    SET_MODEL,
    SET_OPEN,
} from '../actions/detections/types';
import { getPreviousTarget } from '../selectors/detectionSelectors';

export const initialState: IDetectionState = {
    model: null,
    idleTarget: { left: { x: 0, y: 0 }, right: { x: 0, y: 0 } },
    detections: { left: [], right: [] },
    eyesOpenCoefficient: eyelidPosition.OPEN,
    detectionInterval: 0,
    history: [{ left: { x: 0, y: 0 }, right: { x: 0, y: 0 } }],
    colour: { r: 0, g: 0, b: 0 },
};

const detectionActionMapping = {
    [SET_MODEL]: setModel,
    [SET_INTERVAL]: setDetectionInterval,
    [SET_IDLE_TARGET]: setIdleTarget,
    [SET_DETECTIONS]: setDetections,
    [SET_OPEN]: setOpen,
};

const detectionStore = (
    state: IDetectionState = initialState,
    action: DetectionActionType,
): IDetectionState => {
    return action.type in detectionActionMapping
        ? detectionActionMapping[action.type](state, action)
        : state;
};

function setModel(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return { ...state, model: action.payload as IObjectDetector };
}

function setDetectionInterval(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return { ...state, detectionInterval: action.payload as number };
}

function setIdleTarget(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return {
        ...state,
        idleTarget: {
            left: action.payload as ICoords,
            right: action.payload as ICoords,
        },
    };
}

function setDetections(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    const payload = action.payload as ISetDetectionsActionPayload;
    const newHistory = [...state.history];
    if (state.history.length >= maxNumTargetsToConsider) {
        newHistory.shift();
    }
    newHistory.push(payload.previousTarget);

    return {
        ...state,
        detections: payload.detections,
        history: newHistory,
    };
}

function setOpen(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return { ...state, eyesOpenCoefficient: action.payload as number };
}

export default detectionStore;
