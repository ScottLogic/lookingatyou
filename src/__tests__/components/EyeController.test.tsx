import { shallow } from 'enzyme';
import jsdom from 'jsdom';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import {
    EyeController,
    EyeControllerProps,
} from '../../components/eye/EyeController';
import { IConfigState } from '../../store/actions/config/types';
import configStore, {
    initialState as initialConfigState,
} from '../../store/reducers/configReducer';
import { initialState as initialDetectionState } from '../../store/reducers/detectionReducer';
import { IRootStore } from '../../store/reducers/rootReducer';
import { initialState as initialVideoState } from '../../store/reducers/videoReducer';

let props: EyeControllerProps;
let configState: IConfigState;

describe('Eye Controller', () => {
    beforeEach(() => {
        configState = {
            config: {
                xSensitivity: 1,
                ySensitivity: 1,
                fps: 2,
                swapEyes: false,
                toggleDebug: false,
                irisColor: '#ff8080',
                bbox: [0, 0, 0, 0],
            },
        };
        props = {
            width: 500,
            height: 500,
            environment: new jsdom.JSDOM().window,
            target: { left: { x: 250, y: 250 }, right: null },
            config: configState.config,
            dilation: 1,
            openCoefficient: 0.45,
            detected: true,
            videos: [],
        };
    });

    it('should render correctly', () => {
        const mockStore = configureStore();
        const initialState: IRootStore = {
            videoStore: initialVideoState,
            configStore: initialConfigState,
            detectionStore: initialDetectionState,
        };
        const store = mockStore(initialState);
        const wrapper = shallow(
            <Provider store={store}>
                <EyeController {...props} />
            </Provider>,
        ).debug();
        expect(wrapper).toMatchSnapshot();
    });
});
