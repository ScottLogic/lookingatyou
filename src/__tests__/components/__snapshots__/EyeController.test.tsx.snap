// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Eye Controller should render correctly 1`] = `
"<div className=\\"container\\">
  <Eye class=\\"right\\" width={250} height={500} irisColor=\\"#ff8080\\" scleraRadius={111.11111111111111} irisRadius={50} pupilRadius={20.833333333333332} openCoefficient={0.45} dilatedCoefficient={1} innerX={81.78791892748876} innerY={293.21208107251124} fps={2} />
  <Eye class=\\"left\\" width={250} height={500} irisColor=\\"#ff8080\\" scleraRadius={111.11111111111111} irisRadius={50} pupilRadius={20.833333333333332} openCoefficient={0.45} dilatedCoefficient={1} innerX={81.78791892748876} innerY={293.21208107251124} fps={2} />
  <Gradients irisColor=\\"#ff8080\\" />
</div>"
`;
